File: backend/tests/DigitalFamilyCookbook.Tests.Utilities/Mocks/MockRecipe.cs
namespace DigitalFamilyCookbook.Tests.Utilities.Mocks;

public static class MockRecipe
{
    public static RecipeDto GenerateDto() => new RecipeDto
    {
        RecipeId = MockDataGenerator.RandomInteger(),
        Id = MockDataGenerator.RandomId(),
        Name = MockDataGenerator.RandomString(12),
        IsPublic = MockDataGenerator.RandomBoolean(),
        Description = MockDataGenerator.RandomString(500),
        Servings = MockDataGenerator.RandomInteger(1, 8),
        Source = MockDataGenerator.RandomString(20),
        SourceUrl = MockDataGenerator.RandomUrl(),
        Time = MockDataGenerator.RandomInteger(20, 60),
        ActiveTime = MockDataGenerator.RandomInteger(10, 30),
        Calories = MockDataGenerator.RandomInteger(50, 500),
        Protein = MockDataGenerator.RandomInteger(1, 25),
        Carbohydrates = MockDataGenerator.RandomInteger(1, 25),
        Fat = MockDataGenerator.RandomInteger(1, 15),
        Fiber = MockDataGenerator.RandomInteger(1, 15),
        Cholesterol = MockDataGenerator.RandomInteger(1, 25),
        Sugar = MockDataGenerator.RandomInteger(1, 15),
        Ingredients = MockIngredient.GenerateForRecipeDto(5),
        Steps = MockStep.GenerateForRecipeDto(5),
        ImageUrl = $"{Guid.NewGuid().ToString()}_sm.jpg",
        ImageUrlLarge = $"{Guid.NewGuid().ToString()}.jpg",
    };

    public static Recipe GenerateDomainModel() => Recipe.FromDto(GenerateDto());

    public static RecipeApiModel GenerateApiModel() => RecipeApiModel.FromDomainModel(GenerateDomainModel());

    public static List<RecipeDto> GenerateDtoList(int count = 5)
    {
        var recipes = new List<RecipeDto>();

        for (var i = 0; i < count; i += 1)
        {
            recipes.Add(GenerateDto());
        }

        return recipes;
    }

    public static List<Recipe> GenerateDomainModelList(int count = 5)
    {
        var recipes = GenerateDtoList(count);

        return recipes.Select(r => Recipe.FromDto(r)).ToList();
    }

    public static List<RecipeApiModel> GenerateApiModelList(int count = 5)
    {
        var recipes = GenerateDomainModelList(count);

        return recipes.Select(r => RecipeApiModel.FromDomainModel(r)).ToList();
    }
}

File: frontend/src/test/mocks/MockRecipe.ts
import DataGenerator from '@test/DataGenerator';
import Recipe from '@models/Recipe';
import { MockMeatList } from './MockMeat';
import { MockCategoryList } from './MockCategory';
import { MockIngredientList } from './MockIngredient';
import { MockStepList } from './MockStep';
import { MockUserAccount } from './MockUsers';

export const MockRecipe = (): Recipe => {
    const recipeId = DataGenerator.GenerateInteger(1, 100);

    return {
        recipeId,
        id: DataGenerator.GenerateGuid(),
        name: DataGenerator.GenerateString(15),
        description: DataGenerator.GenerateString(100),
        servings: DataGenerator.GenerateInteger(1, 10),
        time: DataGenerator.GenerateInteger(10, 60),
        activeTime: DataGenerator.GenerateInteger(5, 55),
        isPublic: DataGenerator.GenerateInteger(1, 100) % 2 === 0,
        imageUrl: '',
        imageUrlLarge: '',
        imageData: '',
        largeImageData: '',
        calories: DataGenerator.GenerateInteger(0, 400),
        protein: DataGenerator.GenerateInteger(0, 40),
        carbohydrates: DataGenerator.GenerateInteger(0, 40),
        fat: DataGenerator.GenerateInteger(0, 40),
        sugar: DataGenerator.GenerateInteger(0, 40),
        fiber: DataGenerator.GenerateInteger(0, 40),
        cholesterol: DataGenerator.GenerateInteger(0, 40),
        ingredients: MockIngredientList(recipeId),
        steps: MockStepList(recipeId, 5),
        meats: MockMeatList(2),
        categories: MockCategoryList(2),
        notes: [],
        source: DataGenerator.GenerateString(30),
        sourceUrl: DataGenerator.GenerateString(30),
        dateCreated: new Date(),
        dateUpdated: new Date(),
        userAccountId: MockUserAccount.id,
        userAccount: MockUserAccount,
    };
};

export const MockRecipeList = (length = 10): Recipe[] => {
    const recipes: Recipe[] = [];

    for (let i = 0; i < length; i += 1) {
        recipes.push(MockRecipe());
    }

    return recipes;
};
